core/src/main/java/org/elasticsearch/action/admin/indices/flush/TransportShardFlushAction.java

```diff
@@ -61,16 +58,16 @@ protected ActionWriteResponse newResponseInstance() {
     }
 
     @Override
-    protected Tuple<ActionWriteResponse, ShardFlushRequest> shardOperationOnPrimary(ClusterState clusterState, PrimaryOperationRequest shardRequest) throws Throwable {
-        IndexShard indexShard = indicesService.indexServiceSafe(shardRequest.shardId.getIndex()).getShard(shardRequest.shardId.id());
-        indexShard.flush(shardRequest.request.getRequest());
+    protected Tuple<ActionWriteResponse, ShardFlushRequest> shardOperationOnPrimary(MetaData metaData, ShardFlushRequest shardRequest) throws Throwable {
```

Comment:
The only way implementations use the passed in clusterstate to shard operation is to figure out if explicit routing is required (in index and bulk requests), resolve the routing and to know the index creation version from the metadata. IMO, passing in the entire cluster state is redundant whereas passing in the metadata hints at using it to validate the request before performing any operations. I might be missing something though. Do we really need to pass the entire cluster state here?